miio/gateway.py

```diff
@@ -529,45 +580,261 @@ def set_light(self, color_name, brightness):
+        """Return the battery level."""
+        return self._battery
 
-    def get_firmware_version(self) -> Optional[int]:
-        """Returns firmware version"""
+    @command()
+    def send(self, command):
+        """Send a command/query to the subdevice"""
         try:
-            return self.gw.send("get_device_prop", [self.sid, "fw_ver"])[0]
+            return self._gw.send(command, [self.sid])
         except Exception as ex:
-            _LOGGER.debug(
-                "Got an exception while fetching fw_ver: %s", ex, exc_info=True
+            raise GatewayException(
+                "Got an exception while sending command %s" % (command)
+            ) from ex
+
+    @command()
+    def send_arg(self, command, arguments):
+        """Send a command/query including arguments to the subdevice"""
+        try:
+            return self._gw.send(command, arguments, extra_parameters={"sid": self.sid})
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while sending "
+                "command '%s' with arguments '%s'" % (command, str(arguments))
+            ) from ex
+
+    @command(click.argument("property"))
+    def get_property(self, property):
+        """Get the value of a property of the subdevice."""
+        try:
+            response = self._gw.send("get_device_prop", [self.sid, property])
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while fetching property %s" % (property)
+            ) from ex
+
+        if response == []:
+            raise GatewayException(
+                "Empty response while fetching property '%s': %s" % (property, response)
             )
-            return None
 
-    def __repr__(self):
-        return "<Subdevice %s: %s fw: %s bat: %s>" % (
-            self.type,
-            self.sid,
-            self.get_firmware_version(),
-            self.battery(),
-        )
+        return response
+
+    @command(click.argument("properties", nargs=-1))
+    def get_property_exp(self, properties):
+        """Get the value of a bunch of properties of the subdevice."""
+        try:
+            response = self._gw.send(
+                "get_device_prop_exp", [[self.sid] + list(properties)]
+            ).pop()
+        except Exception as ex:
+            raise GatewayException(
+                "Got an exception while fetching properties %s: %s" % (properties)
+            ) from ex
+
+        if len(list(properties)) != len(response):
+            raise GatewayException(
+                "unexpected result while fetching properties %s: %s"
+                % (properties, response)
+            )
+
+        for item in response:
+            if item == '':
+                raise GatewayException(
+                    "One or more empty results while "
+                    "fetching properties %s: %s" % (properties, response)
+                )
```

Comment:
Please remove this. The empty value can very well be a valid value, or it may just a placeholder for invalid state or whatever depending on the device. It is up to `SubDevice` subdevices to handle it accordingly, if raising an exception is wanted behavior.

Reply:
Also, please update the PR description to describe briefly the changes, so that it will be easier for anyone reading the changelog/merged PRs to get an idea.

Reply:
I have removed it